config/initializers/foreman.rb

```diff
@@ -56,3 +56,14 @@
 
   ReportImporter.register_smart_proxy_feature("Puppet")
 end
+
+# Preload here all classes which use Foreman::STI and using registration methods
+# E.g. Base.register_type(BMC)
+# Some constants that use such classes may be defined before all the related classes/models are loaded and registered
+# E.g. InterfaceTypeMapper::ALLOWED_TYPE_NAMES
+Rails.application.reloader.to_prepare do
```

Comment:
This certainly looks cleaner, but what's the difference to `Rails.application.config.to_prepare`? If a plugin provides an implementation to any STI (foreman_discovery's `Host::Discovered` comes to mind), should they do this too and if so, do we have the right hooks in place?

Reply:
I'm note that sure :/ 

TL;DR: I've noticed that `config.to_prepare` blocks run too soon

Currently we do really nasty things (to my taste). For example: Rails during startup runs files under `config/initializers`, where we do some configurations, which is ok. But we for some reason love to call classes that should be loaded as `application` or after basics are done (gem/rails configs). We have `core` and `plugins`, which is like chicken-egg problem. Plugins rely on the core, they don't do anything without it, but they are being loaded before the core due to them being a dependency, just like any other gem. AFAIU, to make the application work, all the deps must be loaded and configured, then the app code is loaded and starts to serve requests. During plugin loading, `config.to_prepare` is being called after all `config/initializers` are done, but before `eager_loading` (which supposedly should load app/**). In these `config.to_prepare` blocks we refer a model or controller before it even got loaded by the application itself, so the constant gets loaded, but without additional context. If eager loading of the app was done earlier, this wouldn't be a problem, since everything under e.g. app/models/nic would be loaded and all the nic types would be registered. Here's an example which makes me feel stupid (or maybe proving it :smile: ) since I'm still not sure what should we do: 
https://github.com/theforeman/foreman_remote_execution/blob/master/lib/foreman_remote_execution/engine.rb#L341

This only line (I'm sure we have many more similar cases) is a headache for me:
 1. Rails/bundler loads plugins before Foreman. At least `to_prepare` blocks from plugins are being run before we run such block in Foreman. In other words: this (https://github.com/theforeman/foreman_remote_execution/blob/master/lib/foreman_remote_execution/engine.rb#L310) is run before this (https://github.com/theforeman/foreman/blob/develop/config/initializers/foreman.rb#L32).
 2. In REX's `to_prepare` block there is invocation of `Api::V2::InterfacesController` class, which gets autoloaded(?) on demand. Since this is the first time the constant is being referenced, it gets parsed and loaded.
 3. In this class there is a reference to another constant -- `InterfaceTypeMapper::ALLOWED_TYPE_NAMES` (https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/interfaces_controller.rb#L37). This makes `InterfaceTypeMapper::ALLOWED_TYPE_NAMES` to be parsed and loaded (since this is the first time it's being referenced).
 4.  In order to load that constant, there is an invocation of another constant -- `Nic::Base` (https://github.com/theforeman/foreman/blob/develop/app/services/interface_type_mapper.rb#L5). The constant is loaded, but `Nic::Base.allowed_types` won't return correct value unless other NIC types are also loaded. Thus, we need to preload these models.
 5. I've tried different ways and places, but got `DEPRECATION WARNING: Initialization autoloaded the constants: ...`. As the message suggested, early referenced constants must be referenced in `Rails.application.reloader.to_prepare` block. Note that this uses `reloader` instead of `config`.
 
 P.S. Currently we load NIC types by calling `require_dependency` at the end of model definitions. But this is a deprecated method, so we shouldn't use these. We could change them to `require`, but what's the point of zeitwerk if we still need that for model files. Rails are aware of STI issues and there are suggestions how to handle it, but none of them work for us: `enable eager loading` - done, but happens too late; `poke db to get types` -- tried, but didn't work in test env.